---
title: "Mendelian randomization study demo"
author: "Yoshihiko Raita, MD MPH MMSc"
date: 
output: html_notebook
---
#Please note that this is still under construction. I need further push of the code. 
Thanks!


#set up the directory
```{r}
#rm(list=ls()); gc();  gc();
#This is my directory. Please set up your directory.
#Please change this to your own directory
setwd("~/Dropbox/seminar_workshop/japan_clinical_epidemiology_2021")
if (!require('dplyr')) install.packages('dplyr'); library('dplyr')
if (!require('readr')) install.packages('readr'); library('readr')
if (!require('stringr')) install.packages('stringr'); library('stringr')
if (!require('data.table')) install.packages('data.table'); library('data.table')
if (!require("remotes")) { install.packages("remotes") } else {}
remotes::install_github("MRCIEU/TwoSampleMR");library('TwoSampleMR')
if (!require('MendelianRandomization')) install.packages('MendelianRandomization'); library('MendelianRandomization')
if (!require('metafor')) install.packages('metafor'); library('metafor')
if (!require("devtools")) { install.packages("devtools") } else {}
devtools::install_github("rondolab/MR-PRESSO");library(MRPRESSO)
if (!require("remotes")) { install.packages("remotes") } else {}
remotes::install_github("CBIIT/LDlinkR");library('LDlinkR')

```




#install GWAS summary statistics of the expsure (Calcium)
In the JAMA paper, they cited the final variants (7 variants) from PLOS Genetics paper. If you would like to select variants from different thresholds, you need to install whole summary statistics data (mostly possible) and selecte variants using a new threshold, but the authors of the PLOS Genetics did not upload a whole summary statistics file. Later, I will explain how to do this briefly by using a different dataset.


```{r}
calcium_exp_data<-read.csv("calcium_exp_variants.csv", header=TRUE)
head(calcium_exp_data,5)
calcium_exp_data$newposition<-paste0(calcium_exp_data$chromosome,":",calcium_exp_data$position)


fstat<-(calcium_exp_data$beta.exposure*calcium_exp_data$beta.exposure)/(calcium_exp_data$se.exposure*calcium_exp_data$se.exposure)
fstat
mean(fstat)
```





#check confounders
download the file
http://csg.sph.umich.edu/willer/public/lipids2013/
```{r}
lipid_gwas_df<-data.table::fread("jointGwasMc_LDL.txt",header = TRUE)
lipid_gwas_df$newposition<-gsub("chr","",lipid_gwas_df$SNP_hg18)
calcium_exp_data%>%left_join(lipid_gwas_df,by="newposition")

```
The P-value of rs780094 is 1.016e-07

```{r}
calcium_exp_data<-calcium_exp_data%>%filter(SNP!="rs780094")
dim(calcium_exp_data)
```




#If you would like to check confoundings associated with life style, you can use UK Biobank
https://broad-ukb-sumstats-us-east-1.s3.amazonaws.com/round2/additive-tsvs/1239.gwas.imputed_v3.both_sexes.tsv.bgz
https://broad-ukb-sumstats-us-east-1.s3.amazonaws.com/round2/annotations/variants.tsv.bgz
annotation_df<- readr::read_tsv(gzfile("variants.tsv.bgz"),col_names=TRUE)
Position of the variant in GRCh37 coordinates.

6*1=6  0.05/6
6*5=30 0.05/30
```{r}
#install.packages("data.table")   
smoking_gwas_df<- readr::read_tsv(gzfile("1239.gwas.imputed_v3.both_sexes.tsv.bgz"),col_names=TRUE)
#smoking_gwas_df<-tidyr::extract(smoking_gwas_df, variant, c('chromosome', 'position', 'allele1', 'allele2'), '(\\d+):(\\d+):(.*):(.*)')
annotation_df<- readr::read_tsv(gzfile("variants.tsv.bgz"),col_names=TRUE)
smoking_gwas_df<-smoking_gwas_df%>%left_join(annotation_df, by="variant")
#smoking_gwas_df$newposition<-paste0(smoking_gwas_df$chr,":",smoking_gwas_df$pos)
merge_calcium_smoke<-calcium_exp_data%>%left_join(smoking_gwas_df, by=c("SNP"="rsid"))
```






# download the whole summary statistics
https://www.ebi.ac.uk/gwas/search?query=26343387
```{r}

cardio_outcome2<-data.table::fread("cad.add.160614.website.txt",header = TRUE)
cardio_outcome2$newposition<-paste0(cardio_outcome2$chr,":",cardio_outcome2$bp_hg19)


```
#Mahanttan plot
You should get the same Manhattan plot
https://gwas.mrcieu.ac.uk/files/ieu-a-7/ieu-a-7_report.html
```{r}
source("https://raw.githubusercontent.com/YinLiLin/CMplot/master/R/CMplot.r")
cardio_outcome2_select<-cardio_outcome2%>%select(markername,chr,bp_hg19,p_dgc)
CMplot(as.data.frame(cardio_outcome2_select),type="p",plot.type="m",LOG10=TRUE,threshold=c(5*10^-8),file="jpg",memo="",dpi=300,cex=c(0.5,0.5,0.5),signal.cex = 0.5, file.output=TRUE,verbose=TRUE,width=14,height=6,chr.labels.angle=45)
```




# Find variants in the outcome file
```{r}
merge_data<-calcium_exp_data%>%inner_join(cardio_outcome2, by=c("SNP"="markername"))
merge_data
merge_data<-merge_data%>%filter(SNP!="rs780094")

merge_data$effect_allele<-str_to_lower(merge_data$effect_allele)
merge_data$noneffect_allele<-str_to_lower(merge_data$noneffect_allele)


write.csv(merge_data, "snps_final_analysis_calcium_cvd.csv", row.names = F)
```
#Final analysis
Harmonization



```{r}

#Create the new file
merge_data_simple<-data.frame(SNP = merge_data$SNP,
                  chr_name=merge_data$chromosome,
                  chrom_start=merge_data$position,
                   beta.exposure =merge_data$beta.exposure,
                  se.exposure = merge_data$se.exposure,
                   effect_allele.exposure = merge_data$effect_allele.exposure,
                  other_allele.exposure =merge_data$other_allele.exposure ,
                  beta.outcome =merge_data$beta,
                  se.outcome = merge_data$se_dgc,
                  effect_allele.outcome = merge_data$effect_allele,
                  other_allele.outcome =merge_data$noneffect_allele,
                   pval.exposure= merge_data$pval.exposure,
                  pval.outcome =merge_data$p_dgc)

lev2 <- unique( c( levels(as.factor(merge_data_simple$effect_allele.outcome)), levels(as.factor(merge_data_simple$effect_allele.exposure)) ) )
merge_data_simple$effect_allele.outcome <- factor(merge_data_simple$effect_allele.outcome, levels=lev2)
merge_data_simple$effect_allele.exposure <- factor(merge_data_simple$effect_allele.exposure, levels=lev2)
merge_data_simple$effect_allele.exposure<-gsub(" ", "",merge_data_simple$effect_allele.exposure, fixed = TRUE)


merge_data_simple$beta.outcome[merge_data_simple$effect_allele.exposure!=merge_data_simple$effect_allele.outcome]<-merge_data_simple$beta.outcome[merge_data_simple$effect_allele.exposure!=merge_data_simple$effect_allele.outcome] * -1


merge_data_simple$fstat<-(merge_data_simple$beta.exposure*merge_data_simple$beta.exposure)/(merge_data_simple$se.exposure*merge_data_simple$se.exposure)

#Take the mean
mean(merge_data_simple$fstat)

#Forest plot with fixed effects, same as Mendelianrandomization IVW with fixed effects 
x<-merge_data_simple$beta.exposure # beta for SNP to exposure
sigmax<-merge_data_simple$se.exposure # its standard errors
y<-merge_data_simple$beta.outcome # beta for SNP to outcome
sigmay<-merge_data_simple$se.outcome # its standard errors
merge_data_simple$Wald<-y/x #Wald estimate
merge_data_simple$Waldvar<-(sigmay^2/x^2) # using Burgess's method
merge_data_simple$lab<-paste(merge_data_simple$SNP, merge_data_simple$gene, sep=" ")

dmres<-rma.uni(yi=merge_data_simple$Wald, vi=merge_data_simple$Waldvar, slab=merge_data_simple$lab, method="FE") #"REML"
forest(dmres, atransf=exp,xlab=" ", mlab="Cardiovascular outcome (OR) per 1 mg/dl increase", at=log(c(.2, 2.5)),xlim=c(-1.7,1.3),cex=.8,showweights = T)


tiff("forest_plot_calcium1.0_CVD.tiff", units="in", width=8, height=5, res=300)
forest(dmres, atransf=exp,xlab=" ", mlab="Cardiovascular outcome (OR) per 1 mg/dl increase", at=log(c(.2, 2.5)),xlim=c(-1.7,1.3),cex=.8,showweights = T)
dev.off()


```




#inverse rank normalization
```{r}
rank_inverse_norm<-function(x){qnorm((rank(x,na.last="keep")-0.5)/sum(!is.na(x)))}

hist(rbeta(10000,5,2))

hist(rank_inverse_norm(rbeta(10000,5,2)))

```



#Unit change example
```{r}
# generate x and y with beta_1=4 and beta_0=10 and i.i.d standard normal errors
set.seed(1)
x<-rnorm(100,30,10)
y<-10+4*x+rnorm(100,0,1)

# estimate basic regression
summary(lm(y~x))


# Now rescale only x
# notice the coefficient and standard errors scale correspondingly
# but the R^2 and t-statistics/p-values do not.
x_rescale<-x*0.5
summary(lm(y~x_rescale))
```





#Analysis by changing units
```{r}


#Create the new file
merge_data_simple<-data.frame(SNP = merge_data$SNP,
                  chr_name=merge_data$chromosome,
                  chrom_start=merge_data$position,
                   beta.exposure =2*merge_data$beta.exposure,
                  se.exposure = 2*merge_data$se.exposure,
                   effect_allele.exposure = merge_data$effect_allele.exposure,
                  other_allele.exposure =merge_data$other_allele.exposure ,
                  beta.outcome =merge_data$beta,
                  se.outcome = merge_data$se_dgc,
                  effect_allele.outcome = merge_data$effect_allele,
                  other_allele.outcome =merge_data$noneffect_allele,
                   pval.exposure= merge_data$pval.exposure,
                  pval.outcome =merge_data$p_dgc)

lev2 <- unique( c( levels(as.factor(merge_data_simple$effect_allele.outcome)), levels(as.factor(merge_data_simple$effect_allele.exposure)) ) )
merge_data_simple$effect_allele.outcome <- factor(merge_data_simple$effect_allele.outcome, levels=lev2)
merge_data_simple$effect_allele.exposure <- factor(merge_data_simple$effect_allele.exposure, levels=lev2)
merge_data_simple$effect_allele.exposure<-gsub(" ", "",merge_data_simple$effect_allele.exposure, fixed = TRUE)


merge_data_simple$beta.outcome[merge_data_simple$effect_allele.exposure!=merge_data_simple$effect_allele.outcome]<-merge_data_simple$beta.outcome[merge_data_simple$effect_allele.exposure!=merge_data_simple$effect_allele.outcome] * -1


merge_data_simple$fstat<-(merge_data_simple$beta.exposure*merge_data_simple$beta.exposure)/(merge_data_simple$se.exposure*merge_data_simple$se.exposure)

#Take the mean
mean(merge_data_simple$fstat)

#Forest plot with fixed effects, same as Mendelianrandomization IVW with fixed effects 
x<-merge_data_simple$beta.exposure # beta for SNP to exposure
sigmax<-merge_data_simple$se.exposure # its standard errors
y<-merge_data_simple$beta.outcome # beta for SNP to outcome
sigmay<-merge_data_simple$se.outcome # its standard errors
merge_data_simple$Wald<-y/x #Wald estimate
merge_data_simple$Waldvar<-(sigmay^2/x^2) # using Burgess's method
merge_data_simple$lab<-paste(merge_data_simple$SNP, merge_data_simple$gene, sep=" ")

dmres<-rma.uni(yi=merge_data_simple$Wald, vi=merge_data_simple$Waldvar, slab=merge_data_simple$lab, method="FE")
forest(dmres, atransf=exp,xlab=" ", mlab="Cardiovascular outcome (OR) per 0.5-mg/dl (1SD) increase", at=log(c(.5, 4.0)),xlim=c(-1.7,1.3),cex=.8,showweights = T)

tiff("forest_plot_calcium0.5_CVD.tiff", units="in", width=8, height=5, res=300)
forest(dmres, atransf=exp,xlab=" ", mlab="Cardiovascular outcome (OR) per 0.5-mg/dl (1SD) increase", at=log(c(.5, 4.0)),xlim=c(-1.7,1.3),cex=.8,showweights = T)
dev.off()
```



#Assessing heterogeniety
```{r}
dmres
```
#Sensitivity Analysis
```{r}
#Mendelian randomization package withfixed effects
MRInputObject <- mr_input(merge_data_simple$beta.exposure, merge_data_simple$se.exposure, merge_data_simple$beta.outcome, merge_data_simple$se.outcome)
MendelianRandomization::mr_ivw(MRInputObject,model="fixed")
MendelianRandomization::mr_ivw(MRInputObject,model="random")

#MR-Egger and MR_Median
MendelianRandomization::mr_egger(MRInputObject)
MendelianRandomization::mr_median(MRInputObject)


#Mendelian randomization package withfixed effects
mr_loo( MRInputObject,alpha = 0.05)


```






#MR-PRESSO
To estimate the causal effect of the risk factors on the asthma outcome with more than five variants, we used MR Pleiotropy RESidual Sum and Outlier (MR-PRESSO) test. MR-PRESSO test detects violation of the restriction exclusion criterion assumption and provides corrected estimates by removing variants which exhibit pleiotropy. 
```{r}
model_presso<-mr_presso(BetaOutcome="beta.outcome", BetaExposure="beta.exposure",SdOutcome="se.outcome",SdExposure="se.exposure",OUTLIERtest=T, DISTORTIONtest=T,data=merge_data_simple,NbDistribution=1000,SignifThreshold=0.05)

model_presso

exp(model_presso$`Main MR results`[1,3]	)
exp(model_presso$`Main MR results`[1,3]-qnorm(0.975)*model_presso$`Main MR results`[1,4]	)
exp(model_presso$`Main MR results`[1,3]+qnorm(0.975)*model_presso$`Main MR results`[1,4]	)


```





#Using TwoSample MR package
https://mrcieu.github.io/TwoSampleMR/

```{r}
library(TwoSampleMR)

calcium_exp_dat <- read_exposure_data(
    filename ="snps_final_analysis_calcium_cvd.csv",
    sep = ",",
    snp_col = "SNP",
    beta_col = "beta.exposure",
    se_col = "se.exposure",
    eaf="exposure_allele_freq",
    effect_allele_col = "effect_allele.exposure",
    other_allele_col = "other_allele.exposure",
    pval_col = "pval.exposure"
)


calcium_exp_dat$beta.exposure<-2*calcium_exp_dat$beta.exposure
calcium_exp_dat$se.exposure<-2*calcium_exp_dat$se.exposure
chd_out_dat <- extract_outcome_data(
    snps = calcium_exp_dat$SNP,
    outcomes = 'ieu-a-7')




dat <- harmonise_data(
    exposure_dat = calcium_exp_dat ,
    outcome_dat = chd_out_dat
)

res <- mr(dat)
#> Analysing 'ieu-a-2' on 'ieu-a-7'
res
class(res)
#change the results to odds ratio
sapply(res[,7],function(x)exp(x))
run_mr_presso(dat, NbDistribution = 1000, SignifThreshold = 0.05)

```







#IL-6 blockade and cardiovascular risk
http://www.phpc.cam.ac.uk/ceu/proteins/
IL-6R
CHR1 154,377,669-250kb or 154,441,926+250kb


```{r}
il6r_gwas_df<-data.table::fread("IL6R.4139.71.2_chrom_1_meta_final_v1.tsv")

il6r_gwas_df$newposition<-paste0(il6r_gwas_df$chromosome,":",il6r_gwas_df$position)

il6r_gwas_df$p_val<-10^il6r_gwas_df$`log(P)`




```
```{r}
threshold<-log(5*10^-8, 10)
il6r_gwas_select_df<-il6r_gwas_df%>%filter(`log(P)`<threshold)

head(il6r_gwas_select_df)
dim(il6r_gwas_select_df)
summary(il6r_gwas_select_df)
```


#Cis- region
154377669-250000
154441926+250000
```{r}
left<-154377669-250000
right<-154441926+250000
il6r_gwas_select_df_select<-il6r_gwas_select_df%>%filter(position>left & position<right )
head(il6r_gwas_select_df_select,5)
```

#annotation
```{r}
annotation_df<- readr::read_tsv(gzfile("variants.tsv.bgz"),col_names=TRUE)
annotation_df$newposition<-paste0(annotation_df$chr,":",annotation_df$pos)
il6r_gwas_select_df_select2<-il6r_gwas_select_df_select%>%left_join(annotation_df,by="newposition")
head(il6r_gwas_select_df_select2,5)
```


```{r}
il6r_gwas_select_df_select2<-il6r_gwas_select_df_select2%>%
  left_join(cardio_outcome2,by="newposition")
il6r_gwas_select_df_select2%>%filter(is.na(beta))
```
#Pruning
https://rdrr.io/github/MRCIEU/TwoSampleMR/man/clump_data.html
```{r}
il6r_gwas_select_df_select2$pval.exposure<-10^(il6r_gwas_select_df_select2$`log(P)`/100)
prune_r2<-0.1
il6r_gwas_select_df_select2$SNP<-il6r_gwas_select_df_select2$rsid
il6r_gwas_select_df_select2$chr_name<-il6r_gwas_select_df_select2$chromosome
il6r_gwas_select_df_select2$chrom_start<-il6r_gwas_select_df_select2$position
il6r_gwas_select_df_select3<-
  TwoSampleMR::clump_data(il6r_gwas_select_df_select2,clump_r2=prune_r2,clump_kb=10000,pop = "EUR")

```

```{r}
il6r_gwas_select_df_select3%>%filter(is.na(beta))
```
```{r}
il6r_gwas_select_df_select3_snpna<-il6r_gwas_select_df_select3%>%filter(is.na(beta))
il6r_gwas_select_df_select3_snpna$newposition2<-paste0("chr",il6r_gwas_select_df_select3_snpna$newposition)
  my_proxies <- LDproxy(snp =   il6r_gwas_select_df_select3_snpna$newposition2,
                      pop = "EUR", 
                      r2d = "r2", 
                      token = "*****"  #Pleas change this to your token
                     )
my_proxies2<-my_proxies #%>%filter(R2>0.5)
my_proxies2


```


#Final analysis
```{r}

merge_data<-il6r_gwas_select_df_select3%>%filter(!is.na(beta))



merge_data$combine_allele1_allele2<-paste0(merge_data$Allele1,":",merge_data$Allele2)
nrow(merge_data)
merge_data<-merge_data%>%filter(combine_allele1_allele2!="a:t"& combine_allele1_allele2!="t:a"& combine_allele1_allele2!="g:c"& combine_allele1_allele2!="c:g")
nrow(merge_data)






#
merge_data_simple<-data.frame(SNP = merge_data$rsid,
                  chr_name=merge_data$chr.x,
                  chrom_start=merge_data$pos,
                   beta.exposure =merge_data$Effect,
                  se.exposure = merge_data$StdErr,
                  effect_allele.exposure = merge_data$Allele1,
                 other_allele.exposure =merge_data$Allele2 ,
                  beta.outcome =merge_data$beta,
                  se.outcome = merge_data$se_dgc,
                  effect_allele.outcome = merge_data$effect_allele,
                  other_allele.outcome =merge_data$noneffect_allele,
                   pval.exposure= merge_data$pval.exposure,
                  pval.outcome =merge_data$p_dgc)
merge_data_simple
lev2 <- unique( c( levels(as.factor(merge_data_simple$effect_allele.outcome)), 
                   levels(as.factor(merge_data_simple$effect_allele.exposure)) ) )
merge_data_simple$effect_allele.outcome <- factor(merge_data_simple$effect_allele.outcome, levels=lev2)
merge_data_simple$effect_allele.exposure <- factor(merge_data_simple$effect_allele.exposure, levels=lev2)
merge_data_simple$effect_allele.exposure<-gsub(" ", "",merge_data_simple$effect_allele.exposure, fixed = TRUE)

merge_data_simple$effect_allele.outcome<-str_to_lower(merge_data_simple$effect_allele.outcome)
merge_data_simple$other_allele.outcome<-str_to_lower(merge_data_simple$other_allele.outcome)




merge_data_simple$beta.outcome[merge_data_simple$effect_allele.exposure!=merge_data_simple$effect_allele.outcome]<-merge_data_simple$beta.outcome[merge_data_simple$effect_allele.exposure!=merge_data_simple$effect_allele.outcome] * -1


merge_data_simple$fstat<-(merge_data_simple$beta.exposure*merge_data_simple$beta.exposure)/(merge_data_simple$se.exposure*merge_data_simple$se.exposure)

#take the mean
mean(merge_data_simple$fstat)

#Forest plot with fixed effects, same as Mendelianrandomization IVW with fixed effects 
x<-merge_data_simple$beta.exposure # beta for SNP to exposure
sigmax<-merge_data_simple$se.exposure # its standard errors
y<-merge_data_simple$beta.outcome # beta for SNP to outcome
sigmay<-merge_data_simple$se.outcome # its standard errors
merge_data_simple$Wald<-y/x #Wald estimate
merge_data_simple$Waldvar<-(sigmay^2/x^2) # using Burgess's method
merge_data_simple$lab<-paste(merge_data_simple$SNP, merge_data_simple$gene, sep=" ")

dmres<-rma.uni(yi=merge_data_simple$Wald, vi=merge_data_simple$Waldvar, slab=merge_data_simple$lab, method="REML")

forest(dmres, atransf=exp,xlab=" ", 
       mlab="\n Cardiovascular outcome (OR) \n per 1 SD increase of sIL6-R concentration", 
       at=log(c(0.6, 1.2)),xlim=c(-0.6,0.5),cex=.8,showweights=T)

tiff("forest_plot_IL6R_CVD.tiff", units="in", width=8, height=5, res=300)
forest(dmres, atransf=exp,xlab=" ", 
       mlab="\n Cardiovascular outcome (OR) \n per 1 SD increase of sIL6-R concentration", 
       at=log(c(0.6, 1.2)),xlim=c(-0.6,0.5),cex=.8,showweights=T)
dev.off()
```

